File: bank-service/src/main/java/rs/edu/raf/mapper/BankAccountMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.ForeignCurrencyBankAccountCreateDTO;
import rs.edu.raf.dto.BusinessBankAccountCreateDTO;
import rs.edu.raf.dto.CurrentBankAccountCreateDTO;
import rs.edu.raf.dto.BankAccountDTO;
import rs.edu.raf.model.accounts.*;
import rs.edu.raf.repository.CurrencyRepository;

import java.math.BigDecimal;

@Component
@AllArgsConstructor
public class BankAccountMapper {
    private CurrencyMapper currencyMapper;
    private CurrencyRepository currencyRepository;

    public BankAccountDTO bankAccountToBankAccountDTO(BankAccount bankAccount) {
        return new BankAccountDTO(bankAccount.getId(), bankAccount.getAccountNumber().toString(), bankAccount.getOwner(),
                bankAccount.getBalance(), bankAccount.getAvailableBalance(), bankAccount.getCreator(),
                bankAccount.getCreationDate(), currencyMapper.currencyToCurrencyDTO(bankAccount.getCurrency()),
                bankAccount.getActive());
    }

    public ForeignCurrencyBankAccount foreignCurrencyBankAccountCreateDtoToForeignCurrencyBankAccount
            (ForeignCurrencyBankAccountCreateDTO foreignCurrencyBankAccountCreateDTO, Long creator) {
        ForeignCurrencyBankAccount foreignCurrencyBankAccount = new ForeignCurrencyBankAccount();

        foreignCurrencyBankAccount.setMaintenancePrice(foreignCurrencyBankAccountCreateDTO.maintenancePrice());
        foreignCurrencyBankAccount.setOwner(foreignCurrencyBankAccountCreateDTO.owner());
        foreignCurrencyBankAccount.setBalance(BigDecimal.ZERO);
        foreignCurrencyBankAccount.setAvailableBalance(BigDecimal.ZERO);
        foreignCurrencyBankAccount.setCreator(creator);
        foreignCurrencyBankAccount.setCreationDate(System.currentTimeMillis());
        foreignCurrencyBankAccount.setCurrency(currencyRepository.findById(foreignCurrencyBankAccountCreateDTO.currencyId()).orElseThrow());
        foreignCurrencyBankAccount.setActive(true);

        return foreignCurrencyBankAccount;
    }

    public BusinessBankAccount businessBankAccountCreateDtoToBusinessBankAccount
            (BusinessBankAccountCreateDTO businessBankAccountCreateDTO, Long creator) {
        BusinessBankAccount businessBankAccount = new BusinessBankAccount();

        businessBankAccount.setOwner(businessBankAccountCreateDTO.companyId());
        businessBankAccount.setBalance(BigDecimal.ZERO);
        businessBankAccount.setAvailableBalance(BigDecimal.ZERO);
        businessBankAccount.setCreator(creator);
        businessBankAccount.setCreationDate(System.currentTimeMillis());
        businessBankAccount.setCurrency(currencyRepository.findByCode("RSD").orElseThrow());
        businessBankAccount.setActive(true);

        return businessBankAccount;
    }

    public CurrentBankAccount currentBankAccountCreateDTOtoCurrentBankAccount
            (CurrentBankAccountCreateDTO currentBankAccountCreateDTO, Long creator) {
        CurrentBankAccount currentBankAccount = new CurrentBankAccount();

        currentBankAccount.setOwner(currentBankAccountCreateDTO.owner());
        currentBankAccount.setBalance(BigDecimal.ZERO);
        currentBankAccount.setAvailableBalance(BigDecimal.ZERO);
        currentBankAccount.setCreator(creator);
        currentBankAccount.setCreationDate(System.currentTimeMillis());
        currentBankAccount.setCurrency(currencyRepository.findByCode("RSD").orElseThrow());
        currentBankAccount.setActive(true);
        currentBankAccount.setMaintenancePrice(currentBankAccountCreateDTO.maintenancePrice());
        currentBankAccount.setType(CurrentBankAccountType.valueOf(currentBankAccountCreateDTO.accountType()));

        return currentBankAccount;
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/mapper/BankAccountMapper.kt
package org.example.bankservicekotlin.mapper

import org.example.bankservicekotlin.dto.BankAccountDTO
import org.example.bankservicekotlin.dto.ForeignCurrencyBankAccountCreateDTO
import org.example.bankservicekotlin.dto.CurrentBankAccountCreateDTO
import org.example.bankservicekotlin.model.accounts.*
import org.example.bankservicekotlin.repository.CurrencyRepository
import org.springframework.stereotype.Component
import java.math.BigDecimal

@Component
class BankAccountMapper(private val currencyMapper: CurrencyMapper, private val currencyRepository: CurrencyRepository) {
    fun bankAccountToBankAccountDTO(bankAccount: BankAccount): BankAccountDTO {
        return BankAccountDTO(
            bankAccount.id, bankAccount.accountNumber.toString(), bankAccount.owner,
            bankAccount.balance, bankAccount.availableBalance, bankAccount.creator,
            bankAccount.creationDate, currencyMapper.currencyToCurrencyDTO(bankAccount.currency),
            bankAccount.active
        )
    }

    fun foreignCurrencyBankAccountCreateDtoToForeignCurrencyBankAccount(
        foreignCurrencyBankAccountCreateDTO: ForeignCurrencyBankAccountCreateDTO,
        creator: Long
    ): ForeignCurrencyBankAccount {
        val currency = currencyRepository.findById(foreignCurrencyBankAccountCreateDTO.currencyId).orElseThrow()
        return ForeignCurrencyBankAccount(
            null,
            null,
            null,
            BigDecimal.ZERO,
            BigDecimal.ZERO,
            creator,
            System.currentTimeMillis(),
            currency,
            true,
            foreignCurrencyBankAccountCreateDTO.maintenancePrice
        )
    }

    fun businessBankAccountCreateDtoToBusinessBankAccount(creator: Long): BusinessBankAccount {
        val currency = currencyRepository.findByCode("RSD").orElseThrow()
        return BusinessBankAccount(
            null,
            null,
            null,
            BigDecimal.ZERO,
            BigDecimal.ZERO,
            creator,
            System.currentTimeMillis(),
            currency,
            true
        )
    }

    fun currentBankAccountCreateDTOtoCurrentBankAccount(
        currentBankAccountCreateDTO: CurrentBankAccountCreateDTO,
        creator: Long
    ): CurrentBankAccount {
        val currency = currencyRepository.findByCode("RSD").orElseThrow();
        return CurrentBankAccount(
            null,
            null,
            null,
            BigDecimal.ZERO,
            BigDecimal.ZERO,
            creator,
            System.currentTimeMillis(),
            currency,
            true,
            CurrentBankAccountType.valueOf(currentBankAccountCreateDTO.accountType),
            currentBankAccountCreateDTO.maintenancePrice
        )
    }
}